internal/vault/credentials/aws.go

```diff
@@ -0,0 +1,220 @@
+	l.authObj = authObj
+	l.providerNamespace = providerNamespace
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		// If AWSCredsRef is not empty, get the secret and read the creds from
+		// there, use the secret UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err := getSecret(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = credsSecret.UID
+	} else if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
+		// Otherwise if the IRSA ref is not empty, read the service account, and
+		// use the service account UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.IRSAServiceAccount,
+		}
+		irsaServiceAccount, err := getSA(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = irsaServiceAccount.UID
+	} else {
+		// At this point either the node role or the instance profile will be
+		// used for credentials, so just generate a new UID
+		l.uid = uuid.NewUUID()
+	}
+
+	return nil
+}
+
+func (l *AWSCredentialProvider) GetCreds(ctx context.Context, client ctrlclient.Client) (map[string]interface{}, error) {
+	logger := log.FromContext(ctx)
+	credsSecret := &corev1.Secret{}
+	irsaToken := ""
+	var irsaConfig *IRSAConfig
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		var err error
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err = getSecret(ctx, client, key)
+		if err != nil {
+			logger.Error(err, "Failed to get secret", "secret_name", l.authObj.Spec.AWS.AWSCredsRef)
+			return nil, err
+		}
+	}
+	if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.IRSAServiceAccount,
+		}
+		irsaServiceAccount, err := getSA(ctx, client, key)
+		if err != nil {
+			logger.Error(err, "Failed to get IRSA service account", "service_account", l.authObj.Spec.AWS.IRSAServiceAccount)
+			return nil, err
+		}
+		irsaConfig = getIRSAConfig(irsaServiceAccount.Annotations)
+
+		token, err := requestSAToken(ctx, client, irsaServiceAccount, irsaConfig.TokenExpiration, []string{irsaConfig.Audience})
+		if err != nil {
+			logger.Error(err, "Failed to get service account token")
+			return nil, err
+		}
+		irsaToken = token.Status.Token
+	}
+
+	config, err := l.getCredentialsConfig(credsSecret, irsaConfig, irsaToken)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: convert logr to something compatible with hclog for use in the
+	// awsutil functions
+	// config.Logger = ...
+
+	creds, err := config.GenerateCredentialChain()
+	if err != nil {
+		return nil, err
+	}
+
+	headerValue := l.authObj.Spec.AWS.HeaderValue
+
+	loginData, err := awsutil.GenerateLoginData(creds, headerValue, config.Region, config.Logger)
+	if err != nil {
+		return nil, err
+	}
+	loginData["role"] = l.authObj.Spec.AWS.Role
```

Comment:
May want to blow up early if `Spec.AWS.Role` is empty, adding some suspenders to the API validation belt...